File: javascript/src/CucumberExpressionGenerator.ts
import CombinatorialGeneratedExpressionFactory from './CombinatorialGeneratedExpressionFactory.js'
import GeneratedExpression from './GeneratedExpression.js'
import ParameterType from './ParameterType.js'
import ParameterTypeMatcher from './ParameterTypeMatcher.js'

export default class CucumberExpressionGenerator {
  constructor(private readonly parameterTypes: () => Iterable<ParameterType<unknown>>) {}

  public generateExpressions(text: string): readonly GeneratedExpression[] {
    const parameterTypeCombinations: Array<Array<ParameterType<unknown>>> = []
    const parameterTypeMatchers = this.createParameterTypeMatchers(text)
    let expressionTemplate = ''
    let pos = 0
    let counter = 0
    // eslint-disable-next-line no-constant-condition
    while (true) {
      let matchingParameterTypeMatchers = []

      for (const parameterTypeMatcher of parameterTypeMatchers) {
        const advancedParameterTypeMatcher = parameterTypeMatcher.advanceTo(pos)
        if (advancedParameterTypeMatcher.find) {
          matchingParameterTypeMatchers.push(advancedParameterTypeMatcher)
        }
      }

      if (matchingParameterTypeMatchers.length > 0) {
        matchingParameterTypeMatchers = matchingParameterTypeMatchers.sort(
          ParameterTypeMatcher.compare
        )

        // Find all the best parameter type matchers, they are all candidates.
        const bestParameterTypeMatcher = matchingParameterTypeMatchers[0]
        const bestParameterTypeMatchers = matchingParameterTypeMatchers.filter(
          (m) => ParameterTypeMatcher.compare(m, bestParameterTypeMatcher) === 0
        )

        // Build a list of parameter types without duplicates. The reason there
        // might be duplicates is that some parameter types have more than one regexp,
        // which means multiple ParameterTypeMatcher objects will have a reference to the
        // same ParameterType.
        // We're sorting the list so preferential parameter types are listed first.
        // Users are most likely to want these, so they should be listed at the top.
        let parameterTypes = []
        for (const parameterTypeMatcher of bestParameterTypeMatchers) {
          if (parameterTypes.indexOf(parameterTypeMatcher.parameterType) === -1) {
            parameterTypes.push(parameterTypeMatcher.parameterType)
          }
        }
        parameterTypes = parameterTypes.sort(ParameterType.compare)

        parameterTypeCombinations.push(parameterTypes)

        expressionTemplate += escape(text.slice(pos, bestParameterTypeMatcher.start))
        expressionTemplate += `{{${counter++}}}`

        pos = bestParameterTypeMatcher.start + bestParameterTypeMatcher.group.length
      } else {
        break
      }

      if (pos >= text.length) {
        break
      }
    }

    expressionTemplate += escape(text.slice(pos))
    return new CombinatorialGeneratedExpressionFactory(
      expressionTemplate,
      parameterTypeCombinations
    ).generateExpressions()
  }

  private createParameterTypeMatchers(text: string): ParameterTypeMatcher[] {
    let parameterMatchers: ParameterTypeMatcher[] = []
    for (const parameterType of this.parameterTypes()) {
      if (parameterType.useForSnippets) {
        parameterMatchers = parameterMatchers.concat(
          CucumberExpressionGenerator.createParameterTypeMatchers2(parameterType, text)
        )
      }
    }
    return parameterMatchers
  }

  private static createParameterTypeMatchers2(
    parameterType: ParameterType<unknown>,
    text: string
  ): ParameterTypeMatcher[] {
    return parameterType.regexpStrings.map(
      (regexp) => new ParameterTypeMatcher(parameterType, regexp, text)
    )
  }
}

function escape(s: string): string {
  return s.replace(/\(/g, '\\(').replace(/{/g, '\\{').replace(/\//g, '\\/')
}

File: dart/lib/src/cucumber_expression_generator.dart
import 'package:cucumber_expressions/src/combinatorial_generated_expression_factory.dart';
import 'package:cucumber_expressions/src/generated_expression.dart';
import 'package:cucumber_expressions/src/parameter_type.dart';
import 'package:cucumber_expressions/src/parameter_type_matcher.dart';
import 'package:cucumber_expressions/src/parameter_type_registry.dart';

/// Generates candidate Cucumber Expressions for a piece of step text based on
/// the available parameter types.
class CucumberExpressionGenerator {
  /// Creates a generator using the supplied parameter type registry.
  CucumberExpressionGenerator(this._parameterTypeRegistry);

  final ParameterTypeRegistry _parameterTypeRegistry;

  /// Generates candidate expressions that would match [text].
  List<GeneratedExpression> generateExpressions(String text) {
    return generateExpressionsForParameterTypes(
      text,
      () => registeredParameterTypes(_parameterTypeRegistry),
    );
  }
}

/// Generates expressions for an internal subset of parameter types.
List<GeneratedExpression> generateExpressionsForParameterTypes(
  String text,
  Iterable<ParameterType<Object?>> Function() parameterTypes,
) {
  final parameterTypeCombinations = <List<ParameterType<Object?>>>[];
  final parameterTypeMatchers = _createParameterTypeMatchers(
    text,
    parameterTypes,
  );
  var expressionTemplate = '';
  var pos = 0;
  var counter = 0;

  while (true) {
    var matchingParameterTypeMatchers = <ParameterTypeMatcher>[];

    for (final parameterTypeMatcher in parameterTypeMatchers) {
      final advancedParameterTypeMatcher = parameterTypeMatcher.advanceTo(pos);
      if (advancedParameterTypeMatcher.find) {
        matchingParameterTypeMatchers.add(advancedParameterTypeMatcher);
      }
    }

    if (matchingParameterTypeMatchers.isNotEmpty) {
      matchingParameterTypeMatchers = matchingParameterTypeMatchers
        ..sort(ParameterTypeMatcher.compare);

      // Find all the best parameter type matchers, they are all candidates.
      final bestParameterTypeMatcher = matchingParameterTypeMatchers[0];
      final bestParameterTypeMatchers = matchingParameterTypeMatchers
          .where(
            (m) =>
                ParameterTypeMatcher.compare(m, bestParameterTypeMatcher) == 0,
          )
          .toList();

      // Build a list of parameter types without duplicates, sorted so
      // preferential parameter types are listed first.
      var parameterTypes = <ParameterType<Object?>>[];
      for (final parameterTypeMatcher in bestParameterTypeMatchers) {
        if (!parameterTypes.contains(parameterTypeMatcher.parameterType)) {
          parameterTypes.add(parameterTypeMatcher.parameterType);
        }
      }
      parameterTypes = parameterTypes..sort(compareParameterTypes);

      parameterTypeCombinations.add(parameterTypes);

      expressionTemplate +=
          _escapeText(text.substring(pos, bestParameterTypeMatcher.start));
      expressionTemplate += '{{${counter++}}}';

      pos = bestParameterTypeMatcher.start +
          bestParameterTypeMatcher.group.length;
    } else {
      break;
    }

    if (pos >= text.length) {
      break;
    }
  }

  expressionTemplate += _escapeText(text.substring(pos));
  return CombinatorialGeneratedExpressionFactory(
    expressionTemplate,
    parameterTypeCombinations,
  ).generateExpressions();
}

List<ParameterTypeMatcher> _createParameterTypeMatchers(
  String text,
  Iterable<ParameterType<Object?>> Function() parameterTypes,
) {
  var parameterMatchers = <ParameterTypeMatcher>[];
  for (final parameterType in parameterTypes()) {
    if (parameterType.useForSnippets) {
      parameterMatchers = parameterMatchers
        ..addAll(_createParameterTypeMatchers2(parameterType, text));
    }
  }
  return parameterMatchers;
}

List<ParameterTypeMatcher> _createParameterTypeMatchers2(
  ParameterType<Object?> parameterType,
  String text,
) {
  return parameterType.regexpStrings
      .map((regexp) => ParameterTypeMatcher(parameterType, regexp, text))
      .toList();
}

String _escapeText(String s) {
  return s.replaceAll('(', r'\(').replaceAll('{', r'\{').replaceAll('/', r'\/');
}
